### PR TITLE
Refactor client component out of client runtime

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes client component loading when it's exported as commonjs

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`
- [ ] The examples guidelines are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing.md#adding-examples)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ccb0011247a0931b47d8213d55659d5685a1ae0d

`yarn testheadless test/e2e/app-dir/index.test.ts` 
- views dir > should pass props from getServerSideProps in root layout
- views dir > should serve from root
- views dir > should include layouts when no direct parent layout
- views dir > should not include parent when not in parent directory with route in directory
- views dir > should use new root layout when provided
- views dir > should not create new root layout when nested (optional)
- views dir > should include parent document when no direct parent layout
- views dir > should not include parent when not in parent directory
- views dir > should serve nested parent
- views dir > should serve dynamic parameter
- views dir > should include document html and body
- views dir > server components > should not serve .client.js as a path
- views dir > server components > should include client component layout with server component route > should include it server-side
- views dir > server components > should include client component layout with server component route > should include it client-side
- views dir > server components > should serve client component > should serve server-side
- views dir > server components > should serve client component > should serve client-side

<details>
<summary>Expand output</summary>

  ● views dir › should pass props from getServerSideProps in root layout

    expect(received).toBe(expected) // Object.is equality

    Expected: "hello world"
    Received: ""

      35 |     const html = await renderViaHTTP(next.url, '/dashboard')
      36 |     const $ = cheerio.load(html)
    > 37 |     expect($('title').text()).toBe('hello world')
         |                               ^
      38 |   })
      39 |
      40 |   it('should serve from pages', async () => {

      at Object.<anonymous> (e2e/app-dir/index.test.ts:37:31)

  ● views dir › should serve from root

    expect(received).toContain(expected) // indexOf

    Expected substring: "hello from root/dashboard"
    Received string:    "<h1>Router</h1><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-root.js\" async=\"\"></script><script>(self.__next_s=self.__next_s||[]).push([0,\"/dashboard,:R0:\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/dashboard,:R0:\",\"M1:{\\\"id\\\":\\\"./node_modules/next/dist/client/components/app-router.client.js\\\",\\\"name\\\":\\\"default\\\",\\\"chunks\\\":[\\\"pages/dashboard/index/page:pages/dashboard/index/page\\\"]}\\nJ0:[\\\"$\\\",\\\"@1\\\",null,{\\\"initialUrl\\\":\\\"/\\\",\\\"children\\\":\\\"@2\\\"}]\\nJ2:[\\\"$\\\",\\\"html\\\",null,{\\\"className\\\":\\\"this-is-the-document-html\\\",\\\"children\\\":[[\\\"$\\\",\\\"head\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"hello world\\\"}]}],[\\\"$\\\",\\\"body\\\",null,{\\\"className\\\":\\\"this-is-the-document-body\\\",\\\"children\\\":[[\\\"$\\\",\\\"h1\\\",null,{\\\"children\\\":\\\"Dashboard\\\"}],[\\\"$\\\",\\\"p\\\",null,{\\\"children\\\":\\\"hello from\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/dashboard,:R0:\",\" root/dashboard\\\"}]]}]]}]\\n\"])</script>"

      55 |   it('should serve from root', async () => {
      56 |     const html = await renderViaHTTP(next.url, '/dashboard')
    > 57 |     expect(html).toContain('hello from root/dashboard')
         |                  ^
      58 |   })
      59 |
      60 |   it('should include layouts when no direct parent layout', async () => {

      at Object.<anonymous> (e2e/app-dir/index.test.ts:57:18)

  ● views dir › should include layouts when no direct parent layout

    expect(received).toBe(expected) // Object.is equality

    Expected: "Dashboard"
    Received: "Router"

      62 |     const $ = cheerio.load(html)
      63 |     // Should not be nested in dashboard
    > 64 |     expect($('h1').text()).toBe('Dashboard')
         |                            ^
      65 |     // Should include the page text
      66 |     expect($('p').text()).toBe('hello from root/dashboard/integrations')
      67 |   })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:64:28)

  ● views dir › should not include parent when not in parent directory with route in directory

    expect(received).toContain(expected) // indexOf

    Expected substring: "<html>"
    Received string:    "<h1>Router</h1><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-root.js\" async=\"\"></script><script>(self.__next_s=self.__next_s||[]).push([0,\"/dashboard/hello,:R0:\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/dashboard/hello,:R0:\",\"M1:{\\\"id\\\":\\\"./node_modules/next/dist/client/components/app-router.client.js\\\",\\\"name\\\":\\\"default\\\",\\\"chunks\\\":[\\\"pages/dashboard/index/page:pages/dashboard/index/page\\\"]}\\nJ0:[\\\"$\\\",\\\"@1\\\",null,{\\\"initialUrl\\\":\\\"/\\\",\\\"children\\\":[\\\"$\\\",\\\"html\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"head\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"Test\\\"}]}],[\\\"$\\\",\\\"body\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"p\\\",null,{\\\"children\\\":\\\"hello from root/dashboard/rootonly/hello\\\"}]}]]}]}]\\n\"])</script>"

      73 |     // new root has to provide it's own custom root layout or the default
      74 |     // is used instead
    > 75 |     expect(html).toContain('<html>')
         |                  ^
      76 |     expect(html).toContain('<body>')
      77 |     expect($('html').hasClass('this-is-the-document-html')).toBeFalsy()
      78 |     expect($('body').hasClass('this-is-the-document-body')).toBeFalsy()

      at Object.<anonymous> (e2e/app-dir/index.test.ts:75:18)

  ● views dir › should use new root layout when provided

    expect(received).toBeTruthy()

    Received: false

      91 |     // new root has to provide it's own custom root layout or the default
      92 |     // is used instead
    > 93 |     expect($('html').hasClass('this-is-another-document-html')).toBeTruthy()
         |                                                                 ^
      94 |     expect($('body').hasClass('this-is-another-document-body')).toBeTruthy()
      95 |
      96 |     // Should not be nested in dashboard

      at Object.<anonymous> (e2e/app-dir/index.test.ts:93:65)

  ● views dir › should not create new root layout when nested (optional)

    expect(received).toBeTruthy()

    Received: false

      110 |     // new root has to provide it's own custom root layout or the default
      111 |     // is used instead
    > 112 |     expect($('html').hasClass('this-is-the-document-html')).toBeTruthy()
          |                                                             ^
      113 |     expect($('body').hasClass('this-is-the-document-body')).toBeTruthy()
      114 |
      115 |     // Should be nested in dashboard

      at Object.<anonymous> (e2e/app-dir/index.test.ts:112:61)

  ● views dir › should include parent document when no direct parent layout

    expect(received).toBeTruthy()

    Received: false

      127 |     const $ = cheerio.load(html)
      128 |
    > 129 |     expect($('html').hasClass('this-is-the-document-html')).toBeTruthy()
          |                                                             ^
      130 |     expect($('body').hasClass('this-is-the-document-body')).toBeTruthy()
      131 |   })
      132 |

      at Object.<anonymous> (e2e/app-dir/index.test.ts:129:61)

  ● views dir › should not include parent when not in parent directory

    expect(received).toBeFalsy()

    Received: "Router"

      135 |     const $ = cheerio.load(html)
      136 |     // Should not be nested in dashboard
    > 137 |     expect($('h1').text()).toBeFalsy()
          |                            ^
      138 |     // Should include the page text
      139 |     expect($('p').text()).toBe('hello from root/dashboard/changelog')
      140 |   })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:137:28)

  ● views dir › should serve nested parent

    expect(received).toBe(expected) // Object.is equality

    Expected: "Dashboard"
    Received: "Router"

      144 |     const $ = cheerio.load(html)
      145 |     // Should be nested in dashboard
    > 146 |     expect($('h1').text()).toBe('Dashboard')
          |                            ^
      147 |     // Should be nested in deployments
      148 |     expect($('h2').text()).toBe('Deployments hello')
      149 |   })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:146:28)

  ● views dir › should serve dynamic parameter

    expect(received).toBe(expected) // Object.is equality

    Expected: "hello from root/dashboard/deployments/[id]. ID is: 123"
    Received: ""

      153 |     const $ = cheerio.load(html)
      154 |     // Should include the page text with the parameter
    > 155 |     expect($('p').text()).toBe(
          |                           ^
      156 |       'hello from root/dashboard/deployments/[id]. ID is: 123'
      157 |     )
      158 |   })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:155:27)

  ● views dir › should include document html and body

    expect(received).toBeTruthy()

    Received: false

      162 |     const $ = cheerio.load(html)
      163 |
    > 164 |     expect($('html').hasClass('this-is-the-document-html')).toBeTruthy()
          |                                                             ^
      165 |     expect($('body').hasClass('this-is-the-document-body')).toBeTruthy()
      166 |   })
      167 |

      at Object.<anonymous> (e2e/app-dir/index.test.ts:164:61)

  ● views dir › server components › should not serve .client.js as a path

    expect(received).toContain(expected) // indexOf

    Expected substring: "hello from root/should-not-serve-client"
    Received string:    "<h1>Router</h1><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-root.js\" async=\"\"></script><script>(self.__next_s=self.__next_s||[]).push([0,\"/should-not-serve-client,:R0:\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/should-not-serve-client,:R0:\",\"M1:{\\\"id\\\":\\\"./node_modules/next/dist/client/components/app-router.client.js\\\",\\\"name\\\":\\\"default\\\",\\\"chunks\\\":[\\\"pages/dashboard/index/page:pages/dashboard/index/page\\\"]}\\nM3:{\\\"id\\\":\\\"./app/should-not-serve-client/page.client.js\\\",\\\"name\\\":\\\"default\\\",\\\"chunks\\\":[\\\"app/should-not-serve-client/page:app/should-not-serve-client/page\\\"]}\\nJ0:[\\\"$\\\",\\\"@1\\\",null,{\\\"initialUrl\\\":\\\"/\\\",\\\"children\\\":\\\"@2\\\"}]\\nJ2:[\\\"$\\\",\\\"html\\\",null,{\\\"className\\\":\\\"this-is-the-document-html\\\",\\\"children\\\":[[\\\"$\\\",\\\"head\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"hello world\\\"\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/should-not-serve-client,:R0:\",\"}]}],[\\\"$\\\",\\\"body\\\",null,{\\\"className\\\":\\\"this-is-the-document-body\\\",\\\"children\\\":[\\\"$\\\",\\\"@3\\\",null,{\\\"children\\\":null}]}]]}]\\n\"])</script>"

      207 |       // Without .client.js should serve
      208 |       const html = await renderViaHTTP(next.url, '/should-not-serve-client')
    > 209 |       expect(html).toContain('hello from root/should-not-serve-client')
          |                    ^
      210 |
      211 |       // Should not serve `.client`
      212 |       const res = await fetchViaHTTP(

      at Object.<anonymous> (e2e/app-dir/index.test.ts:209:20)

  ● views dir › server components › should serve client component › should serve server-side

    expect(received).toBe(expected) // Object.is equality

    Expected: "hello from root/client-component-route. count: 0"
    Received: ""

      236 |         const html = await renderViaHTTP(next.url, '/client-component-route')
      237 |         const $ = cheerio.load(html)
    > 238 |         expect($('p').text()).toBe(
          |                               ^
      239 |           'hello from root/client-component-route. count: 0'
      240 |         )
      241 |       })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:238:31)

  ● views dir › server components › should serve client component › should serve client-side

    page.waitForSelector: Timeout 30000ms exceeded.
    =========================== logs ===========================
    waiting for selector "p"
    ============================================================

      304 |     return this.chain(() => {
      305 |       return page
    > 306 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      307 |         .then(async (el) => {
      308 |           // it seems selenium waits longer and tests rely on this behavior
      309 |           // so we wait for the load event fire before returning

      at lib/browsers/playwright.ts:306:10

  ● views dir › server components › should include client component layout with server component route › should include it server-side

    expect(received).toBe(expected) // Object.is equality

    Expected: "Client Nested. Count: 0"
    Received: "Router"

      256 |         const $ = cheerio.load(html)
      257 |         // Should not be nested in dashboard
    > 258 |         expect($('h1').text()).toBe('Client Nested. Count: 0')
          |                                ^
      259 |         // Should include the page text
      260 |         expect($('p').text()).toBe('hello from root/client-nested')
      261 |       })

      at Object.<anonymous> (e2e/app-dir/index.test.ts:258:32)

  ● views dir › server components › should include client component layout with server component route › should include it client-side

    page.waitForSelector: Timeout 30000ms exceeded.
    =========================== logs ===========================
    waiting for selector "h1"
    ============================================================

      304 |     return this.chain(() => {
      305 |       return page
    > 306 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      307 |         .then(async (el) => {
      308 |           // it seems selenium waits longer and tests rely on this behavior
      309 |           // so we wait for the load event fire before returning

      at lib/browsers/playwright.ts:306:10

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| buildDuration | 16.1s | 15.8s | -307ms |
| buildDurationCached | 6.5s | 6.4s | -16ms |
| nodeModulesSize | 626 MB | 626 MB | ⚠️ +1 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.73 | 3.728 | 0 |
| / avg req/sec | 670.16 | 670.54 | +0.38 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.169 | 1.178 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2139.3 | 2122.22 | ⚠️ -17.08 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 29.2 kB | 29.2 kB | ✓ |
| webpack-HASH.js gzip | 1.54 kB | 1.54 kB | ✓ |
| Overall change | 73 kB | 73 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| buildDuration | 17.9s | 18.1s | ⚠️ +121ms |
| buildDurationCached | 6.5s | 6.5s | -25ms |
| nodeModulesSize | 626 MB | 626 MB | ⚠️ +1 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.727 | 3.701 | -0.03 |
| / avg req/sec | 670.87 | 675.47 | +4.6 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.165 | 1.16 | -0.01 |
| /error-in-render avg req/sec | 2146.5 | 2154.98 | +8.48 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.7 kB | 42.7 kB | ✓ |
| main-HASH.js gzip | 29.7 kB | 29.7 kB | ✓ |
| webpack-HASH.js gzip | 1.54 kB | 1.54 kB | ✓ |
| Overall change | 74 kB | 74 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 311 B | 311 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.9 kB | 2.9 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.84 kB | 5.84 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.78 kB | 2.78 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.4 kB | 16.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/move-router | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: ccb0011247a0931b47d8213d55659d5685a1ae0d

File: test/e2e/app-dir/index.test.ts [hunk 1]
```diff
@@ -52,11 +52,16 @@ describe('views dir', () => {
...
   })
 
-  it('should serve from root', async () => {
+  it('should serve from app', async () => {
     const html = await renderViaHTTP(next.url, '/dashboard')
     expect(html).toContain('hello from root/dashboard')
   })
 
+  it('should serve /index as separate page', async () => {
+    const html = await renderViaHTTP(next.url, '/dashboard/index')
```
Comment (ijjk) [inline test/e2e/app-dir/index.test.ts:61]: Are we not resolving without the `/index` being in the URL correctly? Seems like this should 404 and only be accessible at `/dashboard`

Comment (timneutkens) [reply test/e2e/app-dir/index.test.ts:61]: This was changed in the RFC. You'll be able to colocate `/dashboard/page.js` and `/dashboard/layout.js`. `/dashboard/index/page.js` would resolve to `/dashboard/index`. Basically removing `index` being special.

**Review by ijjk — APPROVED** [review]

**Pull request merged**